readme.md
# Leboncoin API Wrapper

Allow easy acces to leboncoin api using python

## Installation
```bash
pip install .
```

## Usage
```python
from leboncoin_api_wrapper import Leboncoin

lbc = Leboncoin()
lbc.searchFor("iphone")
lbc.setLimit(10)
lbc.maxPrice(2000)
lbc.setDepartement("tarn")
results = lbc.execute()

for ad in results.ads():
    print(ad.subject, ad.price)
print("\n")

for ad in results.shippable_ads():
    print(ad.subject)
    print(ad.body)
    print("\n")
```

## Contributing
Pull requests are welcome. For major changes, please open an issue first to discuss what you would like to change.

Please make sure to update tests as appropriate.

## License
[MIT](https://choosealicense.com/licenses/mit/)

changelog.md
# Changelog
All notable changes to this project will be documented in this file.

## Unreleased
- Auto category is now set to False by default

test.py
from Payload.Payload import Payload
import json


pl = Payload()
pl.setLimit(1)
pl.maxPrice(200)
pl.setDepartement("tarn")

# Cherche les produits avec google home dans le titre
pl.searchFor("google home", True)  # Mettre à false pour cherche partout ailleurs(description...)
resultat = pl.build()

with open("resultat.json", "w+") as jsonFile:
    jsonFile.write(json.dumps(resultat))


Response/Response.py
from pydantic import BaseModel
from typing import List, Optional


class Attribute(BaseModel):
    key: str
    value: str
    values: List[str]
    value_label: str
    generic: bool
    key_label: str = ""


class Geometry(BaseModel):
    type: str
    coordinates: List[float]


class Feature(BaseModel):
    type: str
    geometry: Geometry

class Location(BaseModel):
    country_id: str
    region_id: str
    region_name: str
    department_id: str
    department_name: str
    city_label: str
    city: str
    zipcode: str
    lat: str
    lng: str
    source: str
    provider: str
    id_shape: bool = False
    feature: Feature
    properties: str = "None"


class Owner(BaseModel):
    store_id: str
    user_id: str
    type: str
    name: str
    no_salesmen: bool

class Options(BaseModel):
    has_option: bool
    booster: bool
    photosup: bool
    urgent: bool
    gallery: bool
    sub_toplist: bool
    
class Image(BaseModel):
    thumb_url: str
    small_url: str
    nb_images: int
    urls: List[str]
    urls_thumb: List[str]
    urls_large: List[str]

class Ad(BaseModel):
    list_id: int
    first_publication_date: str
    index_date: str
    status: str
    category_id: str
    category_name: str
    subject: str
    body: str
    ad_type: str
    url: str
    price: List[int]
    price_calendar: Optional[str] = None
    images: Image
    attributes: List[Attribute]
    location: Location
    owner: Owner
    options: Options
    has_phone: bool

class Response(BaseModel):
    total: int
    total_all: int
    total_private: int
    total_active: int
    total_inactive: int
    total_shippable: int
    max_pages: int
    referrer_id: str
    pivot: str
    ads: List[Ad]
    applied_condition: str

leboncoin_api_wrapper/leboncoin.py
from typing import List

from dataclasses import dataclass

import json
import os
#  from pathlib import Path

import requests
from cloudscraper import create_scraper


@dataclass
class Advertisement:
    """
    This class represent an ad with the following attributes
    """
    list_id: str
    first_publication_date: str
    index_date: str
    status: str
    category_id: str
    category_name: str
    subject: str
    body: str
    ad_type: str
    url: str
    price: List[int]
    price_calendar: str
    images: dict
    attributes: List[dict]
    location: dict
    owner: dict
    options: dict
    has_phone: bool

    # noinspection PyTypeChecker
    def __post_init__(self):
        self.price = self.price[0]


class Results:
    def __init__(self, data):
        self._ = data

    def ads(self):
        """
        :return: A list of ads
        """
        return [Advertisement(**ad) for ad in self._["ads"]]

    def shippable_ads(self):
        """
        :return: Return a list of shippable product
        """
        return [Advertisement(**ad) for ad in self._["ads_shippable"]]

    def pprint(self):
        print(json.dumps(self._, indent=4))

    def print(self):
        print(self._)


# noinspection PyTypeChecker,PyPep8Naming
class Leboncoin:
    def __init__(self):
        self._payload = {
            "limit": 35,
            "limit_alu": 3,
            "filters": {
                "enums": {
                    "ad_type": ["offer"]
                },
                "ranges": {
                    "price": {
                        "min": 0,
                    }
                },
                "location": {

                },
                "keywords": {

                },
                "category": {

                }
            }
        }

        current_path = os.path.dirname(os.path.realpath(__file__))
        with open(current_path + "/Ressources/regions.json", "r") as json_file:
            self.region_data = json.load(json_file)
        with open(current_path + "/Ressources/departements.json", "r") as json_file:
            self.dept_data = json.load(json_file)

    def setLimit(self, limit):
        self._payload["limit"] = int(limit)

    def maxPrice(self, price):
        self._payload["filters"]["ranges"]["price"]["max"] = int(price)

    def minPrice(self, price):
        self._payload["filters"]["ranges"]["price"]["min"] = int(price)

    def setRegion(self, region_name):
        for region in self.region_data:
            if region["channel"] == region_name:
                self._payload["filters"]["location"]["locations"] = [{
                    "locationType": "region",
                    "region_id": region["id"],
                    "label": region["name"]
                }]

    def setDepartement(self, dept_name):
        for dept in self.dept_data:
            if dept["channel"].lower() == dept_name:
                self._payload["filters"]["location"]["locations"] = [{
                    "country_id": "FR",
                    "department_id": str(dept["id"]),
                    "locationType": "department",
                    "region_id": dept["region_id"]
                }]

    @staticmethod
    def _get_category(query):
        url = f"https://api.leboncoin.fr/api/parrot/v1/complete?q={query.replace(' ', '%20')}"
        anti_captcha = create_scraper(browser="chrome")
        res = anti_captcha.get(url).json()
        if res:
            assert isinstance(res, List), f"Unexpected answer received from API: {res!r}"
            return str(res[0]["cat_id"])
        else:
            # No category returned
            return None

    def searchFor(self, query, autoCatgory=True):
        self._payload["filters"]["keywords"]["text"] = query
        if autoCatgory:
            category = self._get_category(query)
            if category:
                self._payload["filters"]["category"]["id"] = str(category)

    def setCategory(self, query):
        self._payload["filters"]["category"]["id"] = self._get_category(query)

    def execute(self):
        r = requests.post(
            url="https://api.leboncoin.fr/finder/search",
            data=json.dumps(self._payload),
            headers={
                'User-Agent': 'Mozilla/5.0 (X11; Linux x86_64; rv:67.0) Gecko/20100101 Firefox/67.0',
                'Accept': '*/*',
                'DNT': '1',
            }
        )
        r.raise_for_status()
        return Results(data=r.json())


Payload/Payload.py
import json
from cloudscraper import create_scraper
import requests

class Payload:
    def __init__(self):
        self._payload = {
            "limit": 35,
            "limit_alu": 3,
            "filters": {
                "enums": {
                    "ad_type": ["offer"]
                },
                "ranges": {
                    "price": {
                        "min": 0,
                    }
                },
                "location": {
                    
                },
                "keywords": {
                    
                },
                "category": {
                    
                }
            }
        }
        
        with open("Ressources/regions.json", "r") as json_file:
            self.region_data = json.load(json_file)
        with open("Ressources/departements.json", "r") as json_file:
            self.dept_data = json.load(json_file)
        
    def setLimit(self, limit):
        self._payload["limit"] = int(limit)
        
    def maxPrice(self, price):
        self._payload["filters"]["ranges"]["price"]["max"] = int(price)
    
    def setRegion(self, region_name):
        for region in self.region_data:
            if region["channel"] == region_name:
                self._payload["filters"]["location"]["locations"] = [{
                    "locationType": "region",
                    "region_id": region["id"],
                    "label": region["name"]
                }]
    
    def setDepartement(self, dept_name):
        for dept in self.dept_data:
            if dept["channel"].lower() == dept_name:
                self._payload["filters"]["location"]["locations"] = [{
                    "country_id": "FR",
                    "department_id": str(dept["id"]),
                    "locationType": "department",
                    "region_id": dept["region_id"]
                }]

    def _get_category(self, query):
        url = f"https://api.leboncoin.fr/api/parrot/v1/complete?q={query.replace(' ', '%20')}"
        requests = create_scraper(browser="chrome")
        return str(requests.get(url).json()[0]["cat_id"])
    
    def searchFor(self, query, autoCatgory=True):
        self._payload["filters"]["keywords"]["text"] = query
        if autoCatgory:
            self._payload["filters"]["category"]["id"] = str(self._get_category(query))
    
    def setCategory(self, query):
        self._payload["filters"]["category"]["id"] = self._get_category(query)
        
    def build(self):
        r = requests.post(
            url="https://api.leboncoin.fr/finder/search",
            data=json.dumps(self._payload),
            headers={
                'User-Agent': 'Mozilla/5.0 (X11; Linux x86_64; rv:67.0) Gecko/20100101 Firefox/67.0',
                'Accept': '*/*',
                'DNT': '1',
            }
        )
        if r.status_code != 200:
            raise Exception
        data = r.json()
        return data